File: jvm/src/main/java/edu/kit/kastel/vads/compiler/parser/symbol/Name.java
package edu.kit.kastel.vads.compiler.parser.symbol;

import de.mr_pine.c0ne.lexer.Identifier;
import de.mr_pine.c0ne.lexer.Keyword;

public sealed interface Name permits IdentName, KeywordName {

    static Name forKeyword(Keyword keyword) {
        return new KeywordName(keyword.type);
    }

    static Name forIdentifier(Identifier identifier) {
        return new IdentName(identifier.value);
    }

    String asString();
}

File: jvm/src/main/java/de/mr_pine/c0ne/parser/symbol/Name.kt
package de.mr_pine.c0ne.parser.symbol

import de.mr_pine.c0ne.lexer.Identifier
import de.mr_pine.c0ne.lexer.Keyword

interface Name {
    fun asString(): String

    companion object {
        fun forKeyword(keyword: Keyword): Name {
            return KeywordName(keyword.type)
        }

        fun forIdentifier(identifier: Identifier): Name {
            return IdentName(identifier.value)
        }
    }
}
